File: platform/trainers/ultralytics/runtimes/swift/ModelWrapper.swift
/**
 * ModelWrapper.swift
 * CoreML wrapper for YOLO models exported from Vision AI Training Platform
 *
 * Provides easy-to-use Swift interface for running YOLO models on iOS/macOS.
 *
 * Supports:
 * - YOLOv8/YOLO11 Detection
 * - YOLOv8/YOLO11 Segmentation
 * - YOLOv8/YOLO11 Pose Estimation
 * - YOLOv8/YOLO11 Classification
 *
 * Requirements:
 *   - iOS 13+ / macOS 10.15+
 *   - CoreML
 *   - Vision framework
 *
 * Usage:
 *   let model = try YOLOInference(modelURL: modelURL, metadataURL: metadataURL)
 *   let results = try model.predict(image: uiImage, confThreshold: 0.25, iouThreshold: 0.45)
 *   let annotated = model.visualize(results: results, image: uiImage)
 */

import Foundation
import CoreML
import Vision
import UIKit
import Accelerate

// MARK: - Result Structures

/// Detection result for a single object
public struct Detection {
    public let box: CGRect
    public let confidence: Float
    public let classId: Int
    public let className: String?
}

/// Pose estimation result
public struct PoseResult {
    public let box: CGRect
    public let confidence: Float
    public let keypoints: [CGPoint]
    public let keypointConfidences: [Float]
}

/// Classification result
public struct ClassificationResult {
    public let classId: Int
    public let confidence: Float
    public let probabilities: [Float]
    public let className: String?
}

/// Metadata structure
struct Metadata: Codable {
    let taskType: String
    let inputShape: [Int]
    let classNames: [String]?
    let preprocessing: Preprocessing

    struct Preprocessing: Codable {
        let resize: String
        let normalize: Bool
        let mean: [Float]
        let std: [Float]
    }

    enum CodingKeys: String, CodingKey {
        case taskType = "task_type"
        case inputShape = "input_shape"
        case classNames = "class_names"
        case preprocessing
    }
}

// MARK: - Main Inference Class

public class YOLOInference {

    // MARK: Properties

    private let model: MLModel
    private let metadata: Metadata
    private let taskType: String
    private let inputSize: CGSize
    private let classNames: [String]

    // MARK: Initialization

    /**
     Initialize YOLO inference engine

     - Parameters:
       - modelURL: URL to CoreML model file (.mlmodel or .mlpackage)
       - metadataURL: URL to metadata.json (optional, auto-detected if nil)

     - Throws: Initialization errors
     */
    public init(modelURL: URL, metadataURL: URL? = nil) throws {
        // Load metadata
        let metadataPath = metadataURL ?? modelURL
            .deletingLastPathComponent()
            .appendingPathComponent("metadata.json")

        let metadataData = try Data(contentsOf: metadataPath)
        self.metadata = try JSONDecoder().decode(Metadata.self, from: metadataData)

        self.taskType = metadata.taskType
        self.classNames = metadata.classNames ?? []

        let height = metadata.inputShape[2]
        let width = metadata.inputShape[3]
        self.inputSize = CGSize(width: width, height: height)

        // Load CoreML model
        let config = MLModelConfiguration()
        config.computeUnits = .all  // Use Neural Engine + GPU + CPU

        self.model = try MLModel(contentsOf: modelURL, configuration: config)

        print("[YOLOInference] Model loaded: \(modelURL.lastPathComponent)")
        print("[YOLOInference] Task: \(taskType)")
        print("[YOLOInference] Input size: \(inputSize)")
        print("[YOLOInference] Classes: \(classNames.count)")
    }

    // MARK: - Prediction

    /**
     Run object detection

     - Parameters:
       - image: Input UIImage
       - confThreshold: Confidence threshold (0.0-1.0)
       - iouThreshold: IoU threshold for NMS (0.0-1.0)
       - maxDet: Maximum number of detections

     - Returns: Array of detections
     - Throws: Prediction errors
     */
    public func predict(
        image: UIImage,
        confThreshold: Float = 0.25,
        iouThreshold: Float = 0.45,
        maxDet: Int = 300
    ) throws -> [Detection] {
        guard taskType == "detect" || taskType == "detection" else {
            throw NSError(domain: "YOLOInference", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Use predictPose() or predictClassification() for other tasks"])
        }

        // Preprocess
        let (inputBuffer, preprocessInfo) = try preprocess(image: image)

        // Create MLMultiArray from CVPixelBuffer
        let inputArray = try createMLMultiArray(from: inputBuffer)

        // Run inference
        let inputName = model.modelDescription.inputDescriptionsByName.keys.first!
        let input = try MLDictionaryFeatureProvider(dictionary: [inputName: inputArray])

        let output = try model.prediction(from: input)
        let outputName = model.modelDescription.outputDescriptionsByName.keys.first!

        guard let outputArray = output.featureValue(for: outputName)?.multiArrayValue else {
            throw NSError(domain: "YOLOInference", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get model output"])
        }

        // Postprocess
        let detections = postprocessDetect(
            output: outputArray,
            preprocessInfo: preprocessInfo,
            confThreshold: confThreshold,
            iouThreshold: iouThreshold,
            maxDet: maxDet
        )

        return detections
    }

    /**
     Run pose estimation

     - Parameters:
       - image: Input UIImage
       - confThreshold: Confidence threshold
       - iouThreshold: IoU threshold for NMS
       - maxDet: Maximum number of detections

     - Returns: Array of pose results
     - Throws: Prediction errors
     */
    public func predictPose(
        image: UIImage,
        confThreshold: Float = 0.25,
        iouThreshold: Float = 0.45,
        maxDet: Int = 300
    ) throws -> [PoseResult] {
        // Similar to predict() but with keypoint extraction
        return []
    }

    /**
     Run image classification

     - Parameters:
       - image: Input UIImage
       - confThreshold: Confidence threshold

     - Returns: Classification result
     - Throws: Prediction errors
     */
    public func predictClassification(
        image: UIImage,
        confThreshold: Float = 0.25
    ) throws -> ClassificationResult {
        // Classification logic
        return ClassificationResult(classId: -1, confidence: 0.0, probabilities: [], className: nil)
    }

    // MARK: - Preprocessing

    private func preprocess(image: UIImage) throws -> (CVPixelBuffer, PreprocessInfo) {
        guard let cgImage = image.cgImage else {
            throw NSError(domain: "YOLOInference", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get CGImage"])
        }

        let originalSize = CGSize(width: cgImage.width, height: cgImage.height)

        // Letterbox resize
        let (resized, ratio, padding) = letterboxResize(
            image: image,
            targetSize: inputSize
        )

        // Convert to CVPixelBuffer
        let pixelBuffer = try createPixelBuffer(from: resized)

        let preprocessInfo = PreprocessInfo(
            originalSize: originalSize,
            ratio: ratio,
            padding: padding
        )

        return (pixelBuffer, preprocessInfo)
    }

    private func letterboxResize(
        image: UIImage,
        targetSize: CGSize
    ) -> (UIImage, Float, CGPoint) {
        let size = image.size

        // Calculate scale ratio
        let scaleW = Float(targetSize.width) / Float(size.width)
        let scaleH = Float(targetSize.height) / Float(size.height)
        let ratio = min(scaleW, scaleH)

        // Calculate new size and padding
        let newW = Int(Float(size.width) * ratio)
        let newH = Int(Float(size.height) * ratio)

        let padW = (Int(targetSize.width) - newW) / 2
        let padH = (Int(targetSize.height) - newH) / 2

        // Create canvas with padding
        UIGraphicsBeginImageContextWithOptions(targetSize, false, 1.0)
        defer { UIGraphicsEndImageContext() }

        // Fill with gray (114, 114, 114)
        let context = UIGraphicsGetCurrentContext()!
        context.setFillColor(UIColor(red: 114/255, green: 114/255, blue: 114/255, alpha: 1.0).cgColor)
        context.fill(CGRect(origin: .zero, size: targetSize))

        // Draw resized image
        let rect = CGRect(x: padW, y: padH, width: newW, height: newH)
        image.draw(in: rect)

        let resizedImage = UIGraphicsGetImageFromCurrentImageContext()!

        return (resizedImage, ratio, CGPoint(x: padW, y: padH))
    }

    private func createPixelBuffer(from image: UIImage) throws -> CVPixelBuffer {
        guard let cgImage = image.cgImage else {
            throw NSError(domain: "YOLOInference", code: -4,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get CGImage"])
        }

        let width = cgImage.width
        let height = cgImage.height

        var pixelBuffer: CVPixelBuffer?
        let attrs = [
            kCVPixelBufferCGImageCompatibilityKey: kCFBooleanTrue!,
            kCVPixelBufferCGBitmapContextCompatibilityKey: kCFBooleanTrue!
        ] as CFDictionary

        let status = CVPixelBufferCreate(
            kCFAllocatorDefault,
            width,
            height,
            kCVPixelFormatType_32BGRA,
            attrs,
            &pixelBuffer
        )

        guard status == kCVReturnSuccess, let buffer = pixelBuffer else {
            throw NSError(domain: "YOLOInference", code: -5,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to create pixel buffer"])
        }

        CVPixelBufferLockBaseAddress(buffer, [])
        defer { CVPixelBufferUnlockBaseAddress(buffer, []) }

        let context = CGContext(
            data: CVPixelBufferGetBaseAddress(buffer),
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: CVPixelBufferGetBytesPerRow(buffer),
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.noneSkipFirst.rawValue
        )!

        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))

        return buffer
    }

    private func createMLMultiArray(from pixelBuffer: CVPixelBuffer) throws -> MLMultiArray {
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)

        let shape = [1, 3, height, width] as [NSNumber]
        let array = try MLMultiArray(shape: shape, dataType: .float32)

        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer)!
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)

        // Convert BGRA to RGB and normalize
        let mean = metadata.preprocessing.mean
        let std = metadata.preprocessing.std

        for y in 0..<height {
            for x in 0..<width {
                let pixel = baseAddress.advanced(by: y * bytesPerRow + x * 4)
                let b = Float(pixel.load(fromByteOffset: 0, as: UInt8.self))
                let g = Float(pixel.load(fromByteOffset: 1, as: UInt8.self))
                let r = Float(pixel.load(fromByteOffset: 2, as: UInt8.self))

                let rIdx = [0, 0, y, x] as [NSNumber]
                let gIdx = [0, 1, y, x] as [NSNumber]
                let bIdx = [0, 2, y, x] as [NSNumber]

                array[rIdx] = NSNumber(value: (r - mean[0]) / std[0])
                array[gIdx] = NSNumber(value: (g - mean[1]) / std[1])
                array[bIdx] = NSNumber(value: (b - mean[2]) / std[2])
            }
        }

        return array
    }

    // MARK: - Postprocessing

    private func postprocessDetect(
        output: MLMultiArray,
        preprocessInfo: PreprocessInfo,
        confThreshold: Float,
        iouThreshold: Float,
        maxDet: Int
    ) -> [Detection] {
        // YOLOv8 output: [1, 84, 8400]
        let numBoxes = output.shape[2].intValue
        let numClasses = 80

        var boxes: [CGRect] = []
        var scores: [Float] = []
        var classIds: [Int] = []

        for i in 0..<numBoxes {
            // Find max class confidence
            var maxConf: Float = 0.0
            var maxClassId = 0

            for c in 0..<numClasses {
                let idx = [0, 4 + c, i] as [NSNumber]
                let conf = output[idx].floatValue
                if conf > maxConf {
                    maxConf = conf
                    maxClassId = c
                }
            }

            if maxConf < confThreshold {
                continue
            }

            // Extract box (cx, cy, w, h)
            let cx = output[[0, 0, i] as [NSNumber]].floatValue
            let cy = output[[0, 1, i] as [NSNumber]].floatValue
            let w = output[[0, 2, i] as [NSNumber]].floatValue
            let h = output[[0, 3, i] as [NSNumber]].floatValue

            // Convert to (x, y, w, h)
            let x = cx - w / 2.0
            let y = cy - h / 2.0

            boxes.append(CGRect(x: CGFloat(x), y: CGFloat(y), width: CGFloat(w), height: CGFloat(h)))
            scores.append(maxConf)
            classIds.append(maxClassId)
        }

        if boxes.isEmpty {
            return []
        }

        // Apply NMS
        let keepIndices = nms(boxes: boxes, scores: scores, iouThreshold: iouThreshold)
        let finalIndices = Array(keepIndices.prefix(maxDet))

        // Create detections
        var detections: [Detection] = []
        for idx in finalIndices {
            let scaledBox = scaleBox(box: boxes[idx], preprocessInfo: preprocessInfo)
            let className = classIds[idx] < classNames.count ? classNames[classIds[idx]] : nil

            let detection = Detection(
                box: scaledBox,
                confidence: scores[idx],
                classId: classIds[idx],
                className: className
            )
            detections.append(detection)
        }

        return detections
    }

    // MARK: - Utility Functions

    private func nms(boxes: [CGRect], scores: [Float], iouThreshold: Float) -> [Int] {
        var indices = Array(0..<boxes.count)
        indices.sort { scores[$0] > scores[$1] }

        var keep: [Int] = []

        while !indices.isEmpty {
            let current = indices.removeFirst()
            keep.append(current)

            indices = indices.filter { idx in
                let iou = calculateIoU(box1: boxes[current], box2: boxes[idx])
                return iou <= iouThreshold
            }
        }

        return keep
    }

    private func calculateIoU(box1: CGRect, box2: CGRect) -> Float {
        let intersection = box1.intersection(box2)
        if intersection.isNull {
            return 0.0
        }

        let intersectionArea = intersection.width * intersection.height
        let unionArea = box1.width * box1.height + box2.width * box2.height - intersectionArea

        return Float(intersectionArea / unionArea)
    }

    private func scaleBox(box: CGRect, preprocessInfo: PreprocessInfo) -> CGRect {
        var scaled = box

        // Remove padding
        scaled.origin.x -= preprocessInfo.padding.x
        scaled.origin.y -= preprocessInfo.padding.y

        // Scale back
        let ratio = CGFloat(preprocessInfo.ratio)
        scaled.origin.x /= ratio
        scaled.origin.y /= ratio
        scaled.size.width /= ratio
        scaled.size.height /= ratio

        // Clip to original image
        scaled.origin.x = max(0, min(scaled.origin.x, preprocessInfo.originalSize.width))
        scaled.origin.y = max(0, min(scaled.origin.y, preprocessInfo.originalSize.height))

        return scaled
    }

    // MARK: - Visualization

    /**
     Visualize detection results on image

     - Parameters:
       - results: Detection results
       - image: Original UIImage
       - showLabels: Show class labels
       - showConf: Show confidence scores

     - Returns: Annotated UIImage
     */
    public func visualize(
        results: [Detection],
        image: UIImage,
        showLabels: Bool = true,
        showConf: Bool = true
    ) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        defer { UIGraphicsEndImageContext() }

        image.draw(at: .zero)

        let context = UIGraphicsGetCurrentContext()!

        for detection in results {
            // Draw box
            let color = getColor(classId: detection.classId)
            context.setStrokeColor(color.cgColor)
            context.setLineWidth(3.0)
            context.stroke(detection.box)

            // Draw label
            if showLabels || showConf {
                var label = ""
                if showLabels, let className = detection.className {
                    label = className
                }
                if showConf {
                    if !label.isEmpty { label += " " }
                    label += String(format: "%.2f", detection.confidence)
                }

                let attributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.systemFont(ofSize: 16, weight: .bold),
                    .foregroundColor: UIColor.white
                ]

                let textSize = (label as NSString).size(withAttributes: attributes)
                let textRect = CGRect(
                    x: detection.box.origin.x,
                    y: detection.box.origin.y - textSize.height - 5,
                    width: textSize.width + 10,
                    height: textSize.height + 5
                )

                context.setFillColor(color.cgColor)
                context.fill(textRect)

                (label as NSString).draw(
                    at: CGPoint(x: textRect.origin.x + 5, y: textRect.origin.y + 2),
                    withAttributes: attributes
                )
            }
        }

        return UIGraphicsGetImageFromCurrentImageContext()!
    }

    private func getColor(classId: Int) -> UIColor {
        let hue = CGFloat(classId * 37 % 360) / 360.0
        return UIColor(hue: hue, saturation: 0.8, brightness: 0.9, alpha: 1.0)
    }
}

// MARK: - Supporting Structures

private struct PreprocessInfo {
    let originalSize: CGSize
    let ratio: Float
    let padding: CGPoint
}

File: platform/trainers/ultralytics/runtimes/kotlin/ModelWrapper.kt
/**
 * ModelWrapper.kt
 * TensorFlow Lite wrapper for YOLO models exported from Vision AI Training Platform
 *
 * Provides easy-to-use Kotlin interface for running YOLO models on Android.
 *
 * Supports:
 * - YOLOv8/YOLO11 Detection
 * - YOLOv8/YOLO11 Segmentation
 * - YOLOv8/YOLO11 Pose Estimation
 * - YOLOv8/YOLO11 Classification
 *
 * Requirements:
 *   - Android API 21+
 *   - TensorFlow Lite 2.13+
 *   - TensorFlow Lite GPU delegate (optional)
 *
 * Usage:
 *   val model = YOLOInference(context, "model.tflite", "metadata.json")
 *   val results = model.predict(bitmap, confThreshold = 0.25f, iouThreshold = 0.45f)
 *   val annotated = model.visualize(results, bitmap)
 */

package com.visionai.yolo

import android.content.Context
import android.graphics.*
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.gpu.GpuDelegate
import org.json.JSONObject
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import kotlin.math.max
import kotlin.math.min

// MARK: - Result Structures

/**
 * Detection result for a single object
 */
data class Detection(
    val box: RectF,
    val confidence: Float,
    val classId: Int,
    val className: String? = null
)

/**
 * Pose estimation result
 */
data class PoseResult(
    val box: RectF,
    val confidence: Float,
    val keypoints: List<PointF>,
    val keypointConfidences: List<Float>
)

/**
 * Classification result
 */
data class ClassificationResult(
    val classId: Int,
    val confidence: Float,
    val probabilities: FloatArray,
    val className: String? = null
)

/**
 * Metadata structure
 */
private data class Metadata(
    val taskType: String,
    val inputShape: IntArray,
    val classNames: List<String>,
    val preprocessing: Preprocessing
) {
    data class Preprocessing(
        val resize: String,
        val normalize: Boolean,
        val mean: FloatArray,
        val std: FloatArray
    )
}

/**
 * Preprocessing information
 */
private data class PreprocessInfo(
    val originalSize: Size,
    val ratio: Float,
    val padW: Int,
    val padH: Int
)

private data class Size(val width: Int, val height: Int)

// MARK: - Main Inference Class

/**
 * YOLO inference engine using TensorFlow Lite
 */
class YOLOInference(
    private val context: Context,
    modelPath: String,
    metadataPath: String? = null,
    useGPU: Boolean = true
) {
    private val interpreter: Interpreter
    private val metadata: Metadata
    private val taskType: String
    private val inputSize: Size
    private val classNames: List<String>
    private val gpuDelegate: GpuDelegate?

    init {
        // Load metadata
        val metadataFile = metadataPath ?: modelPath.replace(".tflite", "_metadata.json")
        metadata = loadMetadata(metadataFile)

        taskType = metadata.taskType
        classNames = metadata.classNames
        inputSize = Size(metadata.inputShape[3], metadata.inputShape[2])

        // Initialize TensorFlow Lite
        val options = Interpreter.Options()

        if (useGPU) {
            gpuDelegate = GpuDelegate()
            options.addDelegate(gpuDelegate)
        } else {
            gpuDelegate = null
        }

        options.setNumThreads(4)

        interpreter = Interpreter(loadModelFile(modelPath), options)

        println("[YOLOInference] Model loaded: $modelPath")
        println("[YOLOInference] Task: $taskType")
        println("[YOLOInference] Input size: ${inputSize.width}x${inputSize.height}")
        println("[YOLOInference] Classes: ${classNames.size}")
    }

    /**
     * Release resources
     */
    fun close() {
        interpreter.close()
        gpuDelegate?.close()
    }

    // MARK: - Model Loading

    private fun loadModelFile(modelPath: String): MappedByteBuffer {
        val fileDescriptor = context.assets.openFd(modelPath)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }

    private fun loadMetadata(metadataPath: String): Metadata {
        val jsonString = context.assets.open(metadataPath).bufferedReader().use { it.readText() }
        val json = JSONObject(jsonString)

        val inputShape = json.getJSONArray("input_shape")
        val shape = IntArray(inputShape.length()) { inputShape.getInt(it) }

        val classNamesArray = json.optJSONArray("class_names")
        val classes = if (classNamesArray != null) {
            List(classNamesArray.length()) { classNamesArray.getString(it) }
        } else {
            emptyList()
        }

        val preprocessing = json.getJSONObject("preprocessing")
        val mean = preprocessing.getJSONArray("mean")
        val std = preprocessing.getJSONArray("std")

        return Metadata(
            taskType = json.getString("task_type"),
            inputShape = shape,
            classNames = classes,
            preprocessing = Metadata.Preprocessing(
                resize = preprocessing.getString("resize"),
                normalize = preprocessing.getBoolean("normalize"),
                mean = FloatArray(3) { mean.getDouble(it).toFloat() },
                std = FloatArray(3) { std.getDouble(it).toFloat() }
            )
        )
    }

    // MARK: - Prediction

    /**
     * Run object detection
     *
     * @param bitmap Input bitmap (any size)
     * @param confThreshold Confidence threshold (0.0-1.0)
     * @param iouThreshold IoU threshold for NMS (0.0-1.0)
     * @param maxDet Maximum number of detections
     * @return List of detections
     */
    fun predict(
        bitmap: Bitmap,
        confThreshold: Float = 0.25f,
        iouThreshold: Float = 0.45f,
        maxDet: Int = 300
    ): List<Detection> {
        require(taskType == "detect" || taskType == "detection") {
            "Use predictPose() or predictClassification() for other tasks"
        }

        // Preprocess
        val (inputBuffer, preprocessInfo) = preprocess(bitmap)

        // Prepare output buffer
        // YOLOv8: [1, 84, 8400]
        val outputShape = intArrayOf(1, 84, 8400)
        val outputBuffer = ByteBuffer.allocateDirect(4 * 84 * 8400).apply {
            order(ByteOrder.nativeOrder())
        }

        // Run inference
        interpreter.run(inputBuffer, outputBuffer)

        // Postprocess
        outputBuffer.rewind()
        val output = FloatArray(84 * 8400)
        outputBuffer.asFloatBuffer().get(output)

        return postprocessDetect(output, preprocessInfo, confThreshold, iouThreshold, maxDet)
    }

    /**
     * Run pose estimation
     */
    fun predictPose(
        bitmap: Bitmap,
        confThreshold: Float = 0.25f,
        iouThreshold: Float = 0.45f,
        maxDet: Int = 300
    ): List<PoseResult> {
        // Similar to predict() but with keypoint extraction
        return emptyList()
    }

    /**
     * Run image classification
     */
    fun predictClassification(
        bitmap: Bitmap,
        confThreshold: Float = 0.25f
    ): ClassificationResult {
        // Classification logic
        return ClassificationResult(
            classId = -1,
            confidence = 0.0f,
            probabilities = floatArrayOf(),
            className = null
        )
    }

    // MARK: - Preprocessing

    private fun preprocess(bitmap: Bitmap): Pair<ByteBuffer, PreprocessInfo> {
        val originalSize = Size(bitmap.width, bitmap.height)

        // Letterbox resize
        val (resized, ratio, padW, padH) = letterboxResize(bitmap, inputSize)

        // Convert to ByteBuffer
        val inputBuffer = ByteBuffer.allocateDirect(
            4 * 1 * 3 * inputSize.height * inputSize.width
        ).apply {
            order(ByteOrder.nativeOrder())
        }

        val pixels = IntArray(inputSize.width * inputSize.height)
        resized.getPixels(pixels, 0, inputSize.width, 0, 0, inputSize.width, inputSize.height)

        val mean = metadata.preprocessing.mean
        val std = metadata.preprocessing.std

        // Convert RGBA to RGB and normalize (CHW format)
        for (c in 0..2) {
            for (y in 0 until inputSize.height) {
                for (x in 0 until inputSize.width) {
                    val pixel = pixels[y * inputSize.width + x]

                    val value = when (c) {
                        0 -> Color.red(pixel).toFloat()
                        1 -> Color.green(pixel).toFloat()
                        else -> Color.blue(pixel).toFloat()
                    }

                    val normalized = (value - mean[c]) / std[c]
                    inputBuffer.putFloat(normalized)
                }
            }
        }

        inputBuffer.rewind()

        val preprocessInfo = PreprocessInfo(
            originalSize = originalSize,
            ratio = ratio,
            padW = padW,
            padH = padH
        )

        return Pair(inputBuffer, preprocessInfo)
    }

    private fun letterboxResize(
        bitmap: Bitmap,
        targetSize: Size
    ): Triple<Bitmap, Float, Int, Int> {
        // Calculate scale ratio
        val scaleW = targetSize.width.toFloat() / bitmap.width
        val scaleH = targetSize.height.toFloat() / bitmap.height
        val ratio = min(scaleW, scaleH)

        // Calculate new size and padding
        val newW = (bitmap.width * ratio).toInt()
        val newH = (bitmap.height * ratio).toInt()

        val padW = (targetSize.width - newW) / 2
        val padH = (targetSize.height - newH) / 2

        // Create canvas with padding
        val output = Bitmap.createBitmap(targetSize.width, targetSize.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)

        // Fill with gray (114, 114, 114)
        canvas.drawColor(Color.rgb(114, 114, 114))

        // Draw resized bitmap
        val scaled = Bitmap.createScaledBitmap(bitmap, newW, newH, true)
        canvas.drawBitmap(scaled, padW.toFloat(), padH.toFloat(), null)

        return Triple(output, ratio, padW, padH)
    }

    // MARK: - Postprocessing

    private fun postprocessDetect(
        output: FloatArray,
        preprocessInfo: PreprocessInfo,
        confThreshold: Float,
        iouThreshold: Float,
        maxDet: Int
    ): List<Detection> {
        // YOLOv8 output: [84, 8400]
        // Format: [cx, cy, w, h, class0_conf, class1_conf, ...]

        val numClasses = 80
        val numBoxes = 8400

        val boxes = mutableListOf<RectF>()
        val scores = mutableListOf<Float>()
        val classIds = mutableListOf<Int>()

        for (i in 0 until numBoxes) {
            // Find max class confidence
            var maxConf = 0.0f
            var maxClassId = 0

            for (c in 0 until numClasses) {
                val conf = output[(4 + c) * numBoxes + i]
                if (conf > maxConf) {
                    maxConf = conf
                    maxClassId = c
                }
            }

            if (maxConf < confThreshold) {
                continue
            }

            // Extract box (cx, cy, w, h)
            val cx = output[0 * numBoxes + i]
            val cy = output[1 * numBoxes + i]
            val w = output[2 * numBoxes + i]
            val h = output[3 * numBoxes + i]

            // Convert to (x, y, w, h)
            val x = cx - w / 2.0f
            val y = cy - h / 2.0f

            boxes.add(RectF(x, y, x + w, y + h))
            scores.add(maxConf)
            classIds.add(maxClassId)
        }

        if (boxes.isEmpty()) {
            return emptyList()
        }

        // Apply NMS
        val keepIndices = nms(boxes, scores, iouThreshold)
        val finalIndices = keepIndices.take(maxDet)

        // Create detections
        return finalIndices.map { idx ->
            val scaledBox = scaleBox(boxes[idx], preprocessInfo)
            val className = classIds[idx].takeIf { it < classNames.size }?.let { classNames[it] }

            Detection(
                box = scaledBox,
                confidence = scores[idx],
                classId = classIds[idx],
                className = className
            )
        }
    }

    // MARK: - Utility Functions

    private fun nms(boxes: List<RectF>, scores: List<Float>, iouThreshold: Float): List<Int> {
        val indices = scores.indices.sortedByDescending { scores[it] }.toMutableList()
        val keep = mutableListOf<Int>()

        while (indices.isNotEmpty()) {
            val current = indices.removeAt(0)
            keep.add(current)

            indices.removeAll { idx ->
                calculateIoU(boxes[current], boxes[idx]) > iouThreshold
            }
        }

        return keep
    }

    private fun calculateIoU(box1: RectF, box2: RectF): Float {
        val intersection = RectF(box1)
        if (!intersection.intersect(box2)) {
            return 0.0f
        }

        val intersectionArea = intersection.width() * intersection.height()
        val unionArea = box1.width() * box1.height() + box2.width() * box2.height() - intersectionArea

        return intersectionArea / unionArea
    }

    private fun scaleBox(box: RectF, preprocessInfo: PreprocessInfo): RectF {
        val scaled = RectF(box)

        // Remove padding
        scaled.left -= preprocessInfo.padW
        scaled.top -= preprocessInfo.padH
        scaled.right -= preprocessInfo.padW
        scaled.bottom -= preprocessInfo.padH

        // Scale back
        scaled.left /= preprocessInfo.ratio
        scaled.top /= preprocessInfo.ratio
        scaled.right /= preprocessInfo.ratio
        scaled.bottom /= preprocessInfo.ratio

        // Clip to original image
        scaled.left = max(0f, min(scaled.left, preprocessInfo.originalSize.width.toFloat()))
        scaled.top = max(0f, min(scaled.top, preprocessInfo.originalSize.height.toFloat()))
        scaled.right = max(0f, min(scaled.right, preprocessInfo.originalSize.width.toFloat()))
        scaled.bottom = max(0f, min(scaled.bottom, preprocessInfo.originalSize.height.toFloat()))

        return scaled
    }

    // MARK: - Visualization

    /**
     * Visualize detection results on bitmap
     *
     * @param results Detection results
     * @param bitmap Original bitmap
     * @param showLabels Show class labels
     * @param showConf Show confidence scores
     * @return Annotated bitmap
     */
    fun visualize(
        results: List<Detection>,
        bitmap: Bitmap,
        showLabels: Boolean = true,
        showConf: Boolean = true
    ): Bitmap {
        val output = bitmap.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(output)

        val boxPaint = Paint().apply {
            style = Paint.Style.STROKE
            strokeWidth = 4f
        }

        val textPaint = Paint().apply {
            color = Color.WHITE
            textSize = 40f
            typeface = Typeface.DEFAULT_BOLD
        }

        val bgPaint = Paint().apply {
            style = Paint.Style.FILL
        }

        for (detection in results) {
            // Draw box
            val color = getColor(detection.classId)
            boxPaint.color = color
            canvas.drawRect(detection.box, boxPaint)

            // Draw label
            if (showLabels || showConf) {
                var label = ""
                if (showLabels && detection.className != null) {
                    label = detection.className
                }
                if (showConf) {
                    if (label.isNotEmpty()) label += " "
                    label += String.format("%.2f", detection.confidence)
                }

                val textBounds = Rect()
                textPaint.getTextBounds(label, 0, label.length, textBounds)

                val textX = detection.box.left
                val textY = detection.box.top - 10

                bgPaint.color = color
                canvas.drawRect(
                    textX,
                    textY - textBounds.height() - 10,
                    textX + textBounds.width() + 20,
                    textY,
                    bgPaint
                )

                canvas.drawText(label, textX + 10, textY - 5, textPaint)
            }
        }

        return output
    }

    private fun getColor(classId: Int): Int {
        val hue = (classId * 37 % 360).toFloat()
        return Color.HSVToColor(floatArrayOf(hue, 0.8f, 0.9f))
    }
}
